From 49e16257f042bf26a18831ec285a7486555c46f1 Mon Sep 17 00:00:00 2001
From: Mika Laitio <mika.laitio@amd.com>
Date: Mon, 30 Mar 2026 22:44:47 -0700
Subject: [PATCH 15/19] Add third-party/elfio

Add third-party/elfio project for therock
to allow profiler tools like rocprofiler-sdk to
use it from standard location instead of each subproject
downloading and building it separately.
their own build phase.

This allows standardizasing to single version of these
dependencies and to have a centralized place to fix
problems.

In addition latest release has been patched for the following gcc build
error:

therock/build/third-party/elfio/build_elfio/
dist/include/elfio/elf_types.hpp:30:20:
error: unknown type name 'uint16_t'
9.5	   30 | using Elf_Half   = uint16_t;

fixes: https://github.com/ROCm/TheRock/issues/3517

Signed-off-by: Mika Laitio <mika.laitio@amd.com>
---
 BUILD_TOPOLOGY.toml                           |  9 ++++-
 ...xplicitly-include-cstdint-for-gcc-15.patch | 34 +++++++++++++++++
 third-party/CMakeLists.txt                    | 11 ++++--
 third-party/elfio/CMakeLists.txt              | 37 +++++++++++++++++++
 third-party/elfio/artifact-elfio.toml         |  4 ++
 5 files changed, 91 insertions(+), 4 deletions(-)
 create mode 100644 patches/third-party/elfio/0001-Explicitly-include-cstdint-for-gcc-15.patch
 create mode 100644 third-party/elfio/CMakeLists.txt
 create mode 100644 third-party/elfio/artifact-elfio.toml

diff --git a/BUILD_TOPOLOGY.toml b/BUILD_TOPOLOGY.toml
index 81009cf5..6738ec99 100644
--- a/BUILD_TOPOLOGY.toml
+++ b/BUILD_TOPOLOGY.toml
@@ -415,6 +415,13 @@ feature_name = "HOST_SUITE_SPARSE"
 feature_group = "HOST_MATH"  # Controlled by THEROCK_ENABLE_HOST_MATH (OFF by default)
 disable_platforms = ["windows"]
 
+[artifacts.elfio]
+artifact_group = "third-party-libs"
+type = "target-neutral"
+artifact_deps = []
+feature_group = "CORE"
+disable_platforms = ["windows"]
+
 [artifacts.fftw3]
 artifact_group = "third-party-libs"
 type = "target-neutral"
@@ -693,7 +700,7 @@ disable_platforms = ["windows"]
 [artifacts.rocprofiler-sdk]
 artifact_group = "profiler-core"
 type = "target-neutral"
-artifact_deps = ["core-runtime", "base", "aqlprofile"]  # rocprofiler-register is in base
+artifact_deps = ["core-runtime", "base", "aqlprofile", "elfio"]  # rocprofiler-register is in base
 feature_name = "ROCPROFV3"
 feature_group = "PROFILER"
 disable_platforms = ["windows"]
diff --git a/patches/third-party/elfio/0001-Explicitly-include-cstdint-for-gcc-15.patch b/patches/third-party/elfio/0001-Explicitly-include-cstdint-for-gcc-15.patch
new file mode 100644
index 00000000..642f1b87
--- /dev/null
+++ b/patches/third-party/elfio/0001-Explicitly-include-cstdint-for-gcc-15.patch
@@ -0,0 +1,34 @@
+From 884905d589a5eb1d33af5d191b11fc34a47e736d Mon Sep 17 00:00:00 2001
+From: Mika Laitio <mika.laitio@amd.com>
+Date: Mon, 30 Mar 2026 22:00:46 -0700
+Subject: [PATCH] Explicitly-include-cstdint for gcc 15
+
+Fixes a following error on gcc 15:
+
+therock/build/third-party/elfio/build_elfio/
+dist/include/elfio/elf_types.hpp:30:20:
+error: unknown type name 'uint16_t'
+
+9.5	   30 | using Elf_Half   = uint16_t;
+
+Signed-off-by: Mika Laitio <mika.laitio@amd.com>
+---
+ elfio/elf_types.hpp | 2 ++
+ 1 file changed, 2 insertions(+)
+
+diff --git a/elfio/elf_types.hpp b/elfio/elf_types.hpp
+index c4f6b47..6859b21 100644
+--- a/elfio/elf_types.hpp
++++ b/elfio/elf_types.hpp
+@@ -23,6 +23,8 @@ THE SOFTWARE.
+ #ifndef ELFTYPES_H
+ #define ELFTYPES_H
+ 
++#include <cstdint>
++
+ #ifdef __cplusplus
+ namespace ELFIO {
+ #endif
+-- 
+2.52.0
+
diff --git a/third-party/CMakeLists.txt b/third-party/CMakeLists.txt
index 9a754a00..d2be1ac2 100644
--- a/third-party/CMakeLists.txt
+++ b/third-party/CMakeLists.txt
@@ -11,18 +11,18 @@ add_custom_target(therock-third-party)
 
 # No-dep third party libraries (alphabetical)
 add_subdirectory(boost)
+add_subdirectory(Catch2)
 add_subdirectory(eigen)
+add_subdirectory(fftw3)
 add_subdirectory(flatbuffers)
 add_subdirectory(fmt)
+add_subdirectory(FunctionalPlus)
 add_subdirectory(googletest)
 add_subdirectory(libdivide)
 add_subdirectory(msgpack-cxx)
 add_subdirectory(nlohmann-json)
 add_subdirectory(simde)
 add_subdirectory(yaml-cpp)
-add_subdirectory(Catch2)
-add_subdirectory(FunctionalPlus)
-add_subdirectory(fftw3)
 
 # frugally-deep depends on eigen, FunctionalPlus and nlohmann-json
 add_subdirectory(frugally-deep)
@@ -32,6 +32,11 @@ if(THEROCK_ENABLE_SPDLOG)
   add_subdirectory(spdlog)
 endif()
 
+# elfio depends on amd-llvm toolchain.
+if(THEROCK_ENABLE_ELFIO)
+  add_subdirectory(elfio)
+endif()
+
 # Host math libraries.
 if(THEROCK_ENABLE_HOST_BLAS)
   add_subdirectory(host-blas)
diff --git a/third-party/elfio/CMakeLists.txt b/third-party/elfio/CMakeLists.txt
new file mode 100644
index 00000000..e17a9935
--- /dev/null
+++ b/third-party/elfio/CMakeLists.txt
@@ -0,0 +1,37 @@
+# Copyright Advanced Micro Devices, Inc.
+# SPDX-License-Identifier: MIT
+
+therock_subproject_fetch(therock-elfio-sources
+  CMAKE_PROJECT
+  # Originally mirrored from: "https://github.com/serge1/ELFIO/releases/download/Release_3.12/elfio-3.12.tar.gz"
+  URL "https://rocm-third-party-deps.s3.us-east-2.amazonaws.com/elfio-3.12.tar.gz"
+  URL_HASH "SHA512=13ec6be1f8ee3d7dc96d4ff141d3c199dacd818470c0d902403215571ce3d998afe9e71a8c64e0e69cfb86a4794752d93e2e406025f4812e499e51b67201e23c"
+  PATCH_COMMAND ${Python3_EXECUTABLE} ${CMAKE_SOURCE_DIR}/build_tools/patch_third_party_source.py ${CMAKE_SOURCE_DIR}/patches/third-party/elfio
+)
+
+therock_cmake_subproject_declare(therock-elfio
+  USE_DIST_AMDGPU_TARGETS
+  EXCLUDE_FROM_ALL
+  NO_MERGE_COMPILE_COMMANDS
+  OUTPUT_ON_FAILURE
+  EXTERNAL_SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/source"
+  COMPILER_TOOLCHAIN
+    amd-llvm
+  CMAKE_ARGS
+    -DELFIO_BUILD_EXAMPLES=OFF
+    -DELFIO_BUILD_TESTS=OFF
+)
+therock_cmake_subproject_provide_package(therock-elfio elfio share/elfio/cmake)
+therock_cmake_subproject_activate(therock-elfio)
+add_dependencies(therock-third-party therock-elfio)
+
+therock_provide_artifact(elfio
+  TARGET_NEUTRAL
+  DESCRIPTOR artifact-elfio.toml
+  COMPONENTS
+    dev
+    doc
+    lib
+  SUBPROJECT_DEPS
+    therock-elfio
+)
diff --git a/third-party/elfio/artifact-elfio.toml b/third-party/elfio/artifact-elfio.toml
new file mode 100644
index 00000000..9bd11c64
--- /dev/null
+++ b/third-party/elfio/artifact-elfio.toml
@@ -0,0 +1,4 @@
+# elfio
+[components.dev."third-party/elfio/stage"]
+[components.doc."third-party/elfio/stage"]
+[components.lib."third-party/elfio/stage"]
-- 
2.52.0

